pkgs/development/tools/godot/default.nix

```diff
@@ -1,108 +1,136 @@
-{ stdenv
-, lib
+{ alsa-lib
+, alsa-plugins
 , fetchFromGitHub
-, scons
-, pkg-config
-, udev
+, freetype
+, installShellFiles
+, lib
+, libGLU
+, libpulseaudio
 , libX11
 , libXcursor
+, libXext
+, libXfixes
+, libXi
 , libXinerama
 , libXrandr
 , libXrender
-, libpulseaudio
-, libXi
-, libXext
-, libXfixes
-, freetype
-, openssl
-, alsa-lib
-, alsa-plugins
 , makeWrapper
-, libGLU
-, zlib
+, openssl
+, pkg-config
+, scons
+, stdenv
+, udev
 , yasm
-, withUdev ? true
+, zlib
 }:
 
-let
-  options = {
-    touch = libXi != null;
-    pulseaudio = false;
-    udev = withUdev;
-  };
-in
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (self: {
   pname = "godot";
   version = "3.5.1";
 
   src = fetchFromGitHub {
     owner = "godotengine";
     repo = "godot";
-    rev = "${version}-stable";
+    rev = "${self.version}-stable";
     sha256 = "sha256-uHwTthyhfeQN0R1XjqZ+kGRa5WcpeQzA/DO9hZk4lvU=";
   };
 
-  nativeBuildInputs = [ pkg-config makeWrapper ];
-  buildInputs = [
+  nativeBuildInputs = [
+    installShellFiles
+    pkg-config
+    makeWrapper
     scons
-    udev
+  ];
+
+  buildInputs = [
+    alsa-lib
+    freetype
+    libGLU
+    libpulseaudio
     libX11
     libXcursor
+    libXext
+    libXfixes
+    libXi
     libXinerama
     libXrandr
     libXrender
-    libXi
-    libXext
-    libXfixes
-    freetype
     openssl
-    alsa-lib
-    libpulseaudio
-    libGLU
-    zlib
+    udev
     yasm
+    zlib
   ];
 
-  patches = [ ./pkg_config_additions.patch ./dont_clobber_environment.patch ];
+  patches = [ ./detect.py.patch ];
 
   enableParallelBuilding = true;
 
-  sconsFlags = "target=release_debug platform=x11";
-  preConfigure = ''
-    sconsFlags+=" ${
-      lib.concatStringsSep " "
-      (lib.mapAttrsToList (k: v: "${k}=${builtins.toJSON v}") options)
-    }"
-  '';
+  buildDescription = "x11 tools";
+  buildPlatform = "x11";
+  shouldBuildTools = true;
+  buildTarget = "release_debug";
 
-  outputs = [ "out" "dev" "man" ];
+  sconsFlags = [
+    "platform=${self.buildPlatform}"
+    "tools=${lib.boolToString self.shouldBuildTools}"
+    "target=${self.buildTarget}"
+    "bits=64"
+    "touch=true"
+    "pulseaudio=true"
+    "udev=true"
```

Comment:
Not sure about touch (looks like it's [enabled by default](https://github.com/godotengine/godot/blob/276ab5be07752582b4341c33fb8f53a7887af87a/platform/linuxbsd/detect.py#L52) anyways?), but maybe these should be behind flags? Not really sure, especially since maybe headless doesn't require X11.

Reply:
Based on the detect.py file you linked, all three are enabled by default, and specifying them here may be redundant...

The main reason I have these here is because:
1) In the case of `touch`, it used to enable it if `libXi != null`, which didn't make any sense as Sandro pointed out, because this should never be null.  Thus this simplifies `true`.
2) In the case of `pulseaudio`, it used to be simply set to `false`.  I think audio didn't work for me when it was like this, so I changed it to true.  (Or maybe I'm misremembering, and audio worked fine without it, but I saw no *reason* for it to be false, because it built and ran fine either way.)  Maybe there are circumstances where it shouldn't be enabled, but I'm not sure what those are or how to detect them here.
3) In the case of `udev`, it used to that they had a `withUdev` attribute in the input attrset, and it had a default value of `true`.  Nowhere in nixpkgs was this ever overridden to be false.  So I simplified it to just be `true`, with the expectation that if someone wants to override it, they can do so with overrideAttrs instead.

It's not particularly convenient to override `sconsFlags` via `overrideAttrs`, because it's just a list of strings that happen to contain key-value pairs, so you can't easily change the value of one of the flags.  I used to have it represented as an attrset that's _converted_ into a list of strings, but Sandro indicated this was too complicated so I simplified it to what it is now.  Should be able to replace the `true`s with `${self.someFlag}` though.

Reply:
Oh, and as for whether they should be disabled for headless ... given that they're on by default and the headless build instructions don't _say_ to disable them, I would guess it's not needed.  Maybe if platform = server, it just ignores these features completely.

Reply:
I just removed these three lines.  `"blah=false"`can always be appended in an overrideAttrs if needed.  Flag attributes can be added in the future to control these if it turns out there's good reason to do so...